File: MATLAB/ridgeextract.m
function [R, Y] = ridgeextract(X, scale, step)          % -*-Matlab-*-
% RIDGEEXTRACT  Single-scale ridge extraction
%
% [R, Y] = ridgeextract(X, scale)
%
% INPUTS
%   X      Input image
%   scale  Variance of Gaussian low-pass filter to apply
%
% OUTPUTS
%   R      Array of extracted ridge segments
%   Y      Image with low-pass filter applie
%
% Single-scale ridge extraction algorithm.  A ridge is defined as a
% point for which the brightness assumes a maximum in the principal
% direction of curvature.  If the (p,q) axes at a given point P are
% defined as lieing parallel to the directions of principal
% curvature, then P is classified as a ridge point if:
%
%    Lq = 0, Lqq < 0, Lqq >= Lpp
%
% or:
%
%    Lp = 0, Lpp < 0, Lpp >= Lqq
%
%
% where Lx is the first derivative in the x direction and Lxx the 2nd
% derivative.  This is the same definition used in (Lindeberg, 1998)
% for multi-scale detection (in fact, the algorithm used here is
% identical to the algorithm described in that paper, except using
% only a single scale level).  The Gaussian low-pass filter and
% differencing filters used are as derived in (Lim and Stiehl, 2003).
%
% Ridge segments are found by considering the square formed by four
% adjacent pixels.  Each of the edges is checked for a sign change in
% Lq.  If a sign change is found, a zero-crossing point on the edge
% is found by linear interpolation.  If the curvature is positive,
% the edge is rejected.  If the square has exactly two edges with
% zero-crossing points, they are connected to define a ridge segment.
%
% The output R is a 2x2xN array, where N is the number of segments
% found.  Each column of the 2x2 matrix for a ridge segment is the
% column vector of one of its two endpoints.
%
% REFERENCES
%
%   Lindeberg, T. Edge detection and ridge detection with automatic
%   scale selection. International Journal of Computer Vision,
%   30(2):117--154, 1998.
%
%   Lim, J.Y. and H.S. Stiehl. A generalized discrete scale-space
%   formulation for 2-D and 3-D signals. Scale Space Methods in
%   Computer Vision:132--147, 2003
%
% See also ridgeplot, ridgedemo.

if (nargin < 3);
  % Very conservative downsampling.
  step = 2^(max(0, floor(log(scale)/log(2)) - 4));
end

DSS_KERNEL = [1/6 2/3 1/6];
DIFF_KERNEL = [1/2 0 -1/2];

n_ridge_lines = 0;
R = zeros(2,2,256);

%%%% Create low pass filter kernel
%%
%% This could be speeded up by using recursive self-convolution.
lpf_kern = DSS_KERNEL;
for t = 1:(3*scale);
  lpf_kern = conv(lpf_kern, DSS_KERNEL);
end


%%%% LPF the image
Y = conv2(lpf_kern, lpf_kern, X, 'same');


%%%% Calculate 1st & 2nd partial differences
Dx = conv2(1, DIFF_KERNEL, Y, 'same');
Dy = conv2(DIFF_KERNEL, 1, Y, 'same');
Dxx = conv2(1, DIFF_KERNEL, Dx, 'same');
Dyy = conv2(DIFF_KERNEL, 1, Dy, 'same');
Dxy = conv2(DIFF_KERNEL, 1, Dx, 'same');

%%%% Calculate metrics (Lq and Lqq)
%%
%% q is the unit vector in the most negative principal direction of
%% curvature. Lq is the component of gradient in that direction, and Lqq
%% is the corresponding curvature.
msize = ceil(size(Y)/step);
Lq = zeros(msize);
Lqq = zeros(msize);

for i = 1:msize(1);
  for j = 1:msize(2);
    si = (i-1)*step + 1;
    sj = (j-1)*step + 1;
    %% Calculate curvatures at (i,j) using the eigenvalues of the
    %% Hessian matrix. We choose the eigenvector corresponding with the
    %% greatest curvature *magnitude* as the direction of interest, but
    %% later we'll check that the eigenvalue is negative (what we care
    %% about are ridges rather than troughs, i.e. we're looking for
    %% maxima of value in the image).
    [V, L] = eig([Dxx(si,sj) Dxy(si,sj); Dxy(si,sj) Dyy(si,sj)]);
    if abs(L(1,1)) > abs(L(2,2)); k = 1; else; k = 2; end
    Lqq(i,j) = L(k,k);

    %% Find component of gradient in principal direction of curvature
    q1 = V(1,k); q2 = V(2,k);
    Lq(i,j) = Dx(si,sj)*q1 + Dy(si,sj)*q2;
  end
end

%%%% Find and print out ridge segments.
%%
%% Consider a square of four pixels:
%%
%%         a --- b
%%         |     |
%%         |     |
%%         c --- d
%%
%% Each of the four edges is checked to see if they contain a zero of
%% Lq. If exactly two edges contain a zero of Lq, *and* neither of those
%% edges has Lqq > 0 for at either end, interpolate the zero crossings
%% in each edge and draw a ridge segment.

edge_defs = [0 0 1 0; 1 0 1 1; 1 1 0 1; 0 1 0 0];

crossings = zeros(2);
for i = 1:(msize(1)-1);
  for j = 1:(msize(2)-1);
    n_edges = 0;
    bad_edges = 0;

    for k = 1:size(edge_defs, 1);
      lq1 = Lq(i+edge_defs(k,1),j+edge_defs(k,2));
      lq2 = Lq(i+edge_defs(k,3),j+edge_defs(k,4));
      lqq1 = Lqq(i+edge_defs(k,1),j+edge_defs(k,2));
      lqq2 = Lqq(i+edge_defs(k,3),j+edge_defs(k,4));

      if lq1*lq2 >= 0;  % Check for zero-crossing.
        continue;
      end

      if n_edges >= 2; % Too many zero-crossings?
        break;
      end

      % Interpolate zero-crossing
      delta = lq1 / (lq1 - lq2);

      if ((1-delta)*lqq1 + delta*lqq2) > 0; % Is it a trough?
        continue;
      end

      % Record zero-crossing point
      n_edges = n_edges + 1;

      edge_start = [i+edge_defs(k,1); j+edge_defs(k,2)];
      edge_end = [i+edge_defs(k,3); j+edge_defs(k,4)];
      crossings(:,n_edges) = step*((1 - delta)*edge_start + delta*edge_end);

    end

    if (n_edges ~= 2) || (bad_edges ~= 0);
      continue;
    end

    n_ridge_lines = n_ridge_lines + 1;
    if n_ridge_lines > size(R,3);
      R = cat(3,R,zeros(size(R)));
    end
    R(:,:,n_ridge_lines) = crossings;
  end
end

R = R(:,:,1:n_ridge_lines);


if nargout == 0;
  clf
  subplot(1,2,1);
  imagesc(Y);
  axis image;

  subplot(1,2,2);
  ridgeplot(R);
  axis image;
  axis([0 size(Y,2) 0 size(Y,1)])
end
end

File: python/ridgeextract.py
from numpy import *
from scipy.ndimage.filters import convolve1d
from collections import deque

def derivative(image, row_order, col_order, dest=None):
    def deriv_filter(order):
        K = array([0.5, 0, -0.5], image.dtype);
        f = array([1], image.dtype);
        for i in range(order):
            f = convolve(f, K)
        return f

    if dest == None:
        dest = empty_like(image)

    row_input = image
    result = None
    if (col_order > 0):
        convolve1d(image, deriv_filter(col_order),
                   output=dest, axis=1, mode='constant', cval=0)
        row_input = dest
        result = dest
    if (row_order > 0):
        convolve1d(row_input, deriv_filter(row_order),
                   output=dest, axis=0, mode='constant', cval=0)
        result = dest

    if result == None: # No filtering, just copy over data
        dest[:,:] = image

    return result

class RidgePointInfo:
    def __init__(self, Dqq, Dxx, Dyy, Dxy, value, row, col):
        self.row = row
        self.col = col
        self.Dqq = Dqq
        self.Dxx = Dxx
        self.Dyy = Dyy
        self.Dxy = Dxy
        self.value = value
class RidgeSegmentInfo:
    def __init__(self, start, end):
        self.start = start
        self.end = end
    def reverse(self):
        p = self.end
        self.end = self.start
        self.start = p
        return self

class RidgeExtraction:

    def __init__(self, image):
        # Calculate directional derivatives
        Dx = derivative(image, 1, 0)
        Dy = derivative(image, 0, 1)
        Dxx = derivative(Dx, 1, 0)
        Dyy = derivative(Dy, 0, 1)
        Dxy = derivative(Dx, 0, 1)

        # Calculate metrics

        Dq = empty_like (image)
        Dqq = empty_like (image)

        for k in ndindex(*image.shape):
            A = array([[Dxx[k], Dxy[k]], [Dxy[k], Dyy[k]]])
            w,v = linalg.eigh(A)

            if abs(w[0]) > abs(w[1]):
                dqq = w[0]
                dq = v[0,0]*Dx[k] + v[1,0]*Dy[k]
            else:
                dqq = w[1]
                dq = v[0,1]*Dx[k] + v[1,1]*Dy[k]

            Dq[k] = dq
            Dqq[k] = dqq

        self.image = image
        self.Dx = Dx
        self.Dxx = Dxx
        self.Dyy = Dyy
        self.Dxy = Dxy
        self.Dq = Dq
        self.Dqq = Dqq

    # Linear interpolation
    def itp(self, delta, a, b):
        result = (1-delta)*a + delta*b
        return result

    # Check whether the edge joining start and end contains a ridge
    # point
    def ridge_on_edge(self,start, end):
        i = start
        j = end
        dqi = self.Dq[i]
        dqj = self.Dq[j]
        if dqi * dqj >= 0:
            return None

        # Interpolate zero-crossing of Dq
        dqqi = self.Dqq[i]
        dqqj = self.Dqq[j]

        x = dqi / (dqi - dqj)

        dqq = self.itp(x,dqqi,dqqj)

        if dqq > 0:
            return None

        # Ridge found!
        return RidgePointInfo(dqq,
                              self.itp(x, self.Dxx[i], self.Dxx[j]),
                              self.itp(x, self.Dyy[i], self.Dyy[j]),
                              self.itp(x, self.Dxy[i], self.Dxy[j]),
                              self.itp(x, self.image[i], self.image[j]),
                              self.itp(x, i[0], j[0]),
                              self.itp(x, i[1], j[1]))

    # Find ridge points
    def ridge_points(self):
        for k in ndindex(self.image.shape[0] - 1, self.image.shape[1] - 1):
            r = self.ridge_on_edge (k, (k[0]+1, k[1]))
            if (r != None):
                yield r
            r = self.ridge_on_edge (k, (k[0],   k[1]+1))
            if (r != None):
                yield r

    # Find ridge segments
    def ridge_segments_cached(self):
        # Set up cache for ridge points in previous row/column
        last_row = [self.ridge_on_edge((0,x),(0,x+1))
                           for x in range(self.image.shape[1]-1)]

        for row in range(self.image.shape[0]-1):
            last_col = self.ridge_on_edge((row,0),(row+1,0))
            for col in range(self.image.shape[1]-1):
                # Check for zeros of Dq along each edge of each square
                # formed from four adjacent pixels.
                edge_points = []

                # First check whether there are cached ridge points for
                # the top and left edges of the current square
                if last_row[col] != None:
                    edge_points.append(last_row[col])
                if last_col != None:
                    edge_points.append(last_col)

                # Now check the right and bottom edges of the square, and
                # cache them
                next_col = self.ridge_on_edge((row,col+1),(row+1,col+1))
                if next_col != None:
                    edge_points.append(next_col)
                last_col = next_col

                next_row = self.ridge_on_edge((row+1,col),(row+1,col+1))
                if next_row != None:
                    edge_points.append(next_row)
                last_row[col] = next_row

                # If exactly two ridge points are found on the edges of
                # this square, we add a ridge segment.
                if len(edge_points) != 2:
                    continue

                yield RidgeSegmentInfo(*edge_points)

    def ridge_segments_uncached(self):
        EDGE_INFO = (((0,0),(0,1)),
                     ((0,0),(1,0)),
                     ((0,1),(1,1)),
                     ((1,0),(1,1)))

        for k in ndindex(self.image.shape[0] - 1, self.image.shape[1] - 1):
            edge_points = []
            for edge in EDGE_INFO:
                i = (k[0] + edge[0][0], k[1] + edge[0][1])
                j = (k[0] + edge[1][0], k[1] + edge[1][1])
                r = self.ridge_on_edge(i,j)
                if r != None:
                    edge_points.append(r)
                    if len(edge_points) > 2:
                        break

            if len(edge_points) == 2:
                yield RidgeSegmentInfo(*edge_points)

    ridge_segments = ridge_segments_cached

